qiita_pet/handlers/download.py

```diff
@@ -37,3 +41,77 @@ def get(self, filepath_id):
                         'attachment; filename=%s' % fname)
 
         self.finish()
+
+
+class DownloadStudyBIOMSHandler(BaseHandler):
+    @authenticated
+    @execute_as_transaction
+    def get(self, study_id):
+        study_id = int(study_id)
+        # Check access to study
+        study_info = study_get_req(study_id, self.current_user.id)
+
+        if study_info['status'] != 'success':
+            raise HTTPError(405, "%s: %s, %s" % (study_info['message'],
+                                                 self.current_user.email,
+                                                 str(study_id)))
+
+        study = Study(study_id)
+        user = self.current_user
+        basedir = get_db_files_base_dir()
+        basedir_len = len(basedir) + 1
+        # loop over artifacts and retrieve those that we have access to
+        to_download = []
+        vfabu = validate_filepath_access_by_user
+        for a in study.artifacts():
+            if a.artifact_type == 'BIOM':
+                to_add = True
+                for i, (fid, path, data_type) in enumerate(a.filepaths):
+                    # validate access only of the first artifact filepath,
+                    # the rest have the same permissions
+                    if (i == 0 and not vfabu(user, fid)):
+                        to_add = False
+                        break
+                    if data_type == 'directory':
+                        # If we have a directory, we actually need to list
+                        # all the files from the directory so NGINX can
+                        # actually download all of them
+                        for dp, _, fps in walk(path):
+                            for fname in fps:
+                                fullpath = join(dp, fname)
+                                spath = fullpath
+                                if fullpath.startswith(basedir):
+                                    spath = fullpath[basedir_len:]
+                                to_download.append((fullpath, spath, spath))
+                    elif path.startswith(basedir):
+                        spath = path[basedir_len:]
+                        to_download.append((path, spath, spath))
+                    else:
+                        to_download.append((path, path, path))
+
+                if to_add:
+                    for pt in a.prep_templates:
+                        qmf = pt.qiime_map_fp
+                        if qmf is not None:
+                            sqmf = qmf
+                            if qmf.startswith(basedir):
+                                sqmf = qmf[basedir_len:]
+                            to_download.append(
+                                (qmf, sqmf, 'mapping_files/%s_mapping_file.txt'
+                                            % a.id))
+
+        # If we don't have nginx, write a file that indicates this
+        all_files = '\n'.join(["- %s /protected/%s %s" % (getsize(fp), sfp, n)
```

Comment:
According the the mod_zip documentation you cannot have any extra lines, so I don't think it is possible to add that information.